src/cpp/ext/filters/census/environment_autodetect.cc

```diff
@@ -0,0 +1,425 @@
+                                           std::move(result));
+                }
+                // If fetching from the MetadataServer failed and we were
+                // assuming a GCE environment, fallback to "global".
+                else if (assuming_gce_) {
+                  assuming_gce_ = false;
+                  resource_.resource_type = "global";
+                }
+                attributes_to_fetch_.erase(it);
+              } else {
+                // This should not happen
+                gpr_log(GPR_ERROR,
+                        "An unexpected attribute was seen from the "
+                        "MetadataServer: %s",
+                        attribute.c_str());
+              }
+              if (attributes_to_fetch_.empty()) {
+                resource = std::move(resource_);
+              }
+            }
+            gpr_mu_lock(mu_poll_);
+            notify_poller_ = true;
+            gpr_mu_unlock(mu_poll_);
+            if (resource.has_value()) {
+              auto on_done = std::move(on_done_);
+              Unref();
+              on_done(std::move(resource).value());
+            }
+          });
+    }
+  }
+
+  const std::string project_id_;
+  grpc_pollset* pollset_ = nullptr;
+  grpc_polling_entity pollent_;
+  gpr_mu* mu_poll_ = nullptr;
+  absl::AnyInvocable<void(EnvironmentAutoDetect::ResourceType)> on_done_;
+  grpc_core::Mutex mu_;
+  bool notify_poller_ = false;
+  absl::flat_hash_map<std::string /* metadata_server_attribute */,
+                      std::string /* resource_attribute */>
+      attributes_to_fetch_ ABSL_GUARDED_BY(mu_);
+  EnvironmentAutoDetect::ResourceType resource_ ABSL_GUARDED_BY(mu_);
+  // This would be true if we are assuming the resource to be GCE. In this case,
+  // there is a chance that it will fail and we should instead just use
+  // "global".
+  bool assuming_gce_ ABSL_GUARDED_BY(mu_) = false;
+};
+
+}  // namespace
+
+EnvironmentAutoDetect& EnvironmentAutoDetect::Create(std::string project_id) {
+  static EnvironmentAutoDetect auto_detector(std::move(project_id));
+  return auto_detector;
+}
+
+EnvironmentAutoDetect& EnvironmentAutoDetect::Get() { return Create(""); }
+
+EnvironmentAutoDetect::EnvironmentAutoDetect(std::string project_id)
+    : project_id_(std::move(project_id)) {
+  GPR_ASSERT(!project_id_.empty());
+  new EnvironmentAutoDetectHelper(
+      project_id_, [this](EnvironmentAutoDetect::ResourceType resource) {
+        std::vector<absl::AnyInvocable<void()>> callbacks;
+        {
+          grpc_core::MutexLock lock(&mu_);
+          resource_ = std::make_unique<EnvironmentAutoDetect::ResourceType>(
+              std::move(resource));
+          callbacks = std::move(callbacks_);
+        }
+        for (auto& callback : callbacks) {
+          callback();
+        }
+      });
+}
+
+void EnvironmentAutoDetect::NotifyOnDone(absl::AnyInvocable<void()> callback) {
+  {
+    grpc_core::ReleasableMutexLock lock(&mu_);
```

Comment:
I generally consider `ReleasableMutexLock` to be a code smell.  And in this case, since you're invoking the callback asynchronously, it looks like you don't actually need to release the lock early in the first place.

Reply:
Done. I was intending to reduce the critical region as much as I could, but the EventEngine::Run call is not so bad that we need to worry about it here.